src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -1,59 +1,319 @@
-import io.iohk.ethereum.ledger.Ledger.{BlockPreparationResult, BlockResult, PC, PR, TxResult}
+import io.iohk.ethereum.ledger.BlockQueue.Leaf
+import io.iohk.ethereum.ledger.Ledger._
+import io.iohk.ethereum.utils.Config.SyncConfig
 import io.iohk.ethereum.utils.{BlockchainConfig, DaoForkConfig, Logger}
-import io.iohk.ethereum.validators.{BlockValidator, SignedTransactionValidator}
-import io.iohk.ethereum.domain.UInt256._
+import io.iohk.ethereum.validators._
 import io.iohk.ethereum.vm._
 import org.spongycastle.util.encoders.Hex
 
 import scala.annotation.tailrec
 
 trait Ledger {
 
-  def executeBlock(block: Block, validators: Validators): Either[BlockExecutionError, Seq[Receipt]]
+  def executeBlock(block: Block): Either[BlockExecutionError, Seq[Receipt]]
 
-  def prepareBlock(block: Block, validators: Validators): BlockPreparationResult
+  def prepareBlock(block: Block): BlockPreparationResult
 
   def simulateTransaction(stx: SignedTransaction, blockHeader: BlockHeader): TxResult
+
+  def importBlock(block: Block): BlockImportResult
+
+  def resolveBranch(headers: Seq[BlockHeader]): BranchResolutionResult
 }
 
 //FIXME: Make Ledger independent of BlockchainImpl, for which it should become independent of WorldStateProxy type
-class LedgerImpl(vm: VM, blockchain: BlockchainImpl, blockchainConfig: BlockchainConfig) extends Ledger with Logger {
+//TODO: EC-313: this has grown a bit large, consider splitting the aspects block import, block exec and TX exec
+// scalastyle:off number.of.methods
+class LedgerImpl(
+    vm: VM,
+    blockchain: BlockchainImpl,
+    blockQueue: BlockQueue,
+    blockchainConfig: BlockchainConfig,
+    validators: Validators)
+  extends Ledger with Logger {
+
+  def this(vm: VM, blockchain: BlockchainImpl, blockchainConfig: BlockchainConfig,
+    syncConfig: SyncConfig, validators: Validators) =
+    this(vm, blockchain, new BlockQueue(blockchain, syncConfig), blockchainConfig, validators)
 
   val blockRewardCalculator = new BlockRewardCalculator(blockchainConfig.monetaryPolicyConfig)
 
-  def executeBlock(
-    block: Block,
-    validators: Validators): Either[BlockExecutionError, Seq[Receipt]] = {
+  // scalastyle:off method.length
+  /**
+    * Tries to import the block as the new best block in the chain or enqueue it for later processing
+    * @param block - block to be imported
+    * @return One of:
+    *         - [[BlockImportedToTop]] - if the block was added as the new best block
+    *         - [[BlockEnqueued]] - block is stored in the [[BlockQueue]]
+    *         - [[ChainReorganised]] - a better new branch was found causing chain reorganisation
+    *         - [[DuplicateBlock]] - block already exists either in the main chain or in the queue
+    *         - [[BlockImportFailed]] - block failed to execute (when importing to top or reorganising the chain)
+    */
+  def importBlock(block: Block): BlockImportResult = {
```

Comment:
Is this block being validated before queuing? If not, is there an chance a malicious peer keep sending us invalid nodes to fill up our memory/queue? I know there is a `cleanUp` by block number but not by queued block count.

_PS: sorry if this doesn't make any sense, I was doing a quick pass over this PR_

Reply:
Good point 🤔

Reply:
Even if we do initial validation, the malicious peer could send us blocks that pass this validation and fail to execute later. Then we're not even able to tell which peer sent us this block and blacklist it.

Unless full PoW would be sufficiently prevent a malicious peer from generating such blocks, @LukasGasior1 ?

Reply:
I need to take a deeper look but what about `io.iohk.ethereum.validators.BlockHeaderValidatorImpl#validatePoW` ?

Reply:
ps: I mean, is it enough?

Reply:
Yep, anyway, I will use it now (it will be later improved), but as we discussed with @LukasGasior1, we'll also need to validate the blocks for the accepted difficulty range (we may not have a parent for enqueued block, but we know it's within certain accepted number range). That should take care of spam attacks.

Apart from that do you still think I should implement a size limit to the `BlockQueue`? If so, which blocks should be discarded when reaching the limit? (This smells like a follow-up task 😉 )

Reply:
Maybe this even this validation should be a follow-up task? Pinging @mcsherrylabs

Reply:
Regarding queue limits, I need to take a second look at the code to answer your question. I haven't gone through reorg logic yet (I'm wondering if it can switch to another branch where might be some missing parents but we know it's the longest chain for some reason).  **I think doing this in a separate task is better tho**

Reply:
@rtkaczyk are you adding validating if difficulty is within required range in this PR?

Reply:
Yes, I am